contracts/external/cw-payroll/src/state.rs

```diff
@@ -0,0 +1,397 @@
+    pub vesting_schedule: Curve,
+    /// The denom of a token (cw20 or native)
+    pub denom: CheckedDenom,
+    /// Title of the payroll item, for example for a bug bounty "Fix issue in contract.rs"
+    pub title: Option<String>,
+    /// Description of the payroll item, a more in depth description of how to meet the payroll conditions
+    pub description: Option<String>,
+    /// The status of the vesting payment
+    pub status: VestingPaymentStatus,
+    /// The amount of the vesting payment that has been staked
+    pub staked_amount: Uint128,
+    /// Info about staked vesting payment rewards
+    pub rewards: VestingPaymentRewards,
+}
+
+impl VestingPayment {
+    /// Create a new VestingPayment from CheckedVestingParams
+    pub fn new(checked_vesting_params: CheckedVestingParams) -> Result<Self, ContractError> {
+        let vesting_payment = Self {
+            status: VestingPaymentStatus::Active,
+            claimed_amount: Uint128::zero(),
+            staked_amount: Uint128::zero(),
+            canceled_at_time: None,
+            recipient: checked_vesting_params.recipient,
+            amount: checked_vesting_params.amount,
+            denom: checked_vesting_params.denom,
+            vesting_schedule: checked_vesting_params.vesting_schedule,
+            title: checked_vesting_params.title,
+            description: checked_vesting_params.description,
+            rewards: VestingPaymentRewards {
+                pending: Decimal::zero(),
+                paid_rewards_per_token: Decimal::zero(),
+            },
+        };
+
+        Ok(vesting_payment)
+    }
+
+    pub fn get_vested_amount_by_seconds(&self, time: u64) -> Result<Uint128, OverflowError> {
+        let vesting_funds = self.vesting_schedule.value(time);
+        self.amount.checked_sub(vesting_funds)
+    }
+
+    pub fn calc_pending_rewards(
+        &mut self,
+        new_rewards_per_token: Decimal,
+        staked: Uint128,
+    ) -> Result<(), ContractError> {
+        // No staked amount, so no rewards
+        if staked.is_zero() {
+            self.rewards.paid_rewards_per_token = new_rewards_per_token;
+            return Ok(());
+        }
+
+        let rewards_per_token_to_pay = new_rewards_per_token - self.rewards.paid_rewards_per_token;
+
+        // We don't need to update anything, nothing to calculate
+        if rewards_per_token_to_pay.is_zero() {
+            return Ok(());
+        }
+
+        self.rewards.pending +=
+            rewards_per_token_to_pay.checked_mul(Decimal::from_atomics(staked, 0)?)?;
+
+        self.rewards.paid_rewards_per_token = new_rewards_per_token;
+
+        Ok(())
+    }
+
+    //// TODO Do we need this?
+    // pub fn reset_pending_rewards(&mut self) {
+    //     self.rewards.pending -= self.rewards.pending.floor();
+    // }
+
+    /// Turn pending decimal to u128 to send tokens
+    pub fn get_pending_rewards(&self) -> Result<Uint128, ContractError> {
+        // TODO: Find a better way of doing this?
```

Comment:
@0xekez would love your thoughts here, and also a sanity check on this math. :joy:

Reply:
We have some similar code in the Mesh Security repo, and I think it's messy.

Reply:
DELETED IT.

Reply:
:heart: